crates/ruff_db/src/source.rs

```diff
@@ -0,0 +1,128 @@
+use std::ops::Deref;
+use std::sync::Arc;
+
+use ruff_source_file::LineIndex;
+
+use crate::vfs::VfsFile;
+use crate::Db;
+
+/// Reads the content of file.
+#[salsa::tracked]
+pub fn source_text(db: &dyn Db, file: VfsFile) -> SourceText {
+    let content = file.read(db);
+
+    SourceText {
+        inner: Arc::from(content),
+    }
+}
+
+/// Computes the [`LineIndex`] for `file`.
+#[salsa::tracked]
+pub fn line_index(db: &dyn Db, file: VfsFile) -> LineIndex {
+    let source = source_text(db, file);
+
+    LineIndex::from_source_text(&source)
+}
+
+/// The source text of a [`VfsFile`].
+///
+/// Cheap cloneable in `O(1)`.
+#[derive(Clone, Eq, PartialEq)]
+pub struct SourceText {
+    inner: Arc<str>,
+}
+
+impl SourceText {
+    pub fn as_str(&self) -> &str {
+        &self.inner
+    }
+}
+
+impl Deref for SourceText {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl std::fmt::Debug for SourceText {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("SourceText").field(&self.inner).finish()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use filetime::FileTime;
+    use salsa::EventKind;
+
+    use ruff_source_file::OneIndexed;
+    use ruff_text_size::TextSize;
+
+    use crate::file_system::FileSystemPath;
+    use crate::source::{line_index, source_text};
+    use crate::tests::TestDb;
+    use crate::Db;
+
+    #[test]
+    fn re_runs_query_when_file_revision_changes() {
+        let mut db = TestDb::new();
+        let path = FileSystemPath::new("test.py");
+
+        db.file_system_mut().write_file(path, "x = 10".to_string());
+
+        let file = db.file(path);
+
+        assert_eq!(&*source_text(&db, file), "x = 10");
+
+        db.file_system_mut().write_file(path, "x = 20".to_string());
+        file.set_revision(&mut db).to(FileTime::now().into());
```

Comment:
It's surprising to me that we have to manually update the revision here; should `write_file` do that itself?

Reply:
The `FileSystem` that I have in mind should be independent of the `vfs` and salsa db. It's just an abstraction over `std::fs`. What I have in mind to avoid the `set_revision` call here is that we have an `apply_changes` method where we pass all file changes (adds, remove, modified) that updates the vfs state. It's probably the `FileSystem'`s or the host's responsibility to collect all made changes.

Reply:
That makes sense, and would address my concern: I don't want to expose/use APIs that make it easy to change a file's contents but forget to bump its revision.

I think this also gets back to what I find confusing about the `Vfs` and `FileSystem` structure, which is that there is no single entity which really encapsulates our virtual file system. Where does this `apply_changes` method live? I guess it is just a free function that takes a `Db` and some changes to apply?

Probably this is just something I have to get used to with Salsa, that the `Db` owns all the state, so most API that we use for managing state is just a function that takes a `Db`.

Reply:
> Where does this apply_changes method live? I guess it is just a free function that takes a Db and some changes to apply?

It would probably be a free standing function that takes `&mut Db`. And yes, the fact that it is a free standing function takes some time to get used to. An alternative is to make it a method in `Vfs`, but I think that won't work because it would require holding a mutable reference to `Db` and a read only reference to `FileSystem`. 

> Probably this is just something I have to get used to with Salsa, that the Db owns all the state, so most API that we use for managing state is just a function that takes a Db.

yes, that takes some time to get used to. It's also something I'm still figuring out